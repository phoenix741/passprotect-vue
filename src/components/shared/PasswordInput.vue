<template lang="pug">
v-text-field(
    :label="label"
    :type="passwordVisibility ? 'text' : 'password'"
    loading
    :required="required"
    :error-messages="errorMessages"

    :append-icon="passwordVisibility ? 'visibility' : 'visibility_off'"
    @click:append="() => (passwordVisibility = !passwordVisibility)"

    :persistent-hint="true"
    :hint="zxcvbn.feedback.warning"

    :value="value"
    @input="onChange($event)")
    v-progress-linear(slot="progress",:value="zxcvbnProgress",:color="zxcvbnColor",height="3")
</template>

<script type="text/babel">
import zxcvbn from 'zxcvbn'

export default {
  props: {
    label: String,
    value: {
      type: String,
      required: false
    },
    words: {
      type: Array,
      default: () => [],
      required: false
    },
    required: {
      type: Boolean,
      default: false
    },
    errorMessages: {
      type: Array,
      default: () => [],
      required: false
    }
  },
  methods: {
    onChange (value) {
      this.$emit('input', value)
    }
  },
  data () {
    return {
      passwordVisibility: false
    }
  },
  computed: {
    zxcvbn () {
      return (this.value && zxcvbn(this.value, this.words.filter(e => !!e))) || { feedback: {} }
    },
    zxcvbnProgress () {
      return this.zxcvbn.score * 25
    },
    zxcvbnColor () {
      return ['red accent-2', 'orange', 'amber', 'light-green ', 'green'][this.zxcvbn.score]
    }
  }
}
</script>
