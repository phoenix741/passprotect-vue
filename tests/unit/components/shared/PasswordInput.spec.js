import { shallowMount } from '@vue/test-utils'
import PasswordInput from '@/components/shared/PasswordInput.vue'

describe('PasswordInput.vue', () => {
  it('Test that the component show a progress bar', () => {
    const PasswordInputComponent = shallowMount(PasswordInput, {
      propsData: {
        id: 'passwordInput',
        name: 'passwordInputName',
        label: 'Mon mot de passe',
        value: 'test'
      }
    })
    expect(PasswordInputComponent.element).toMatchSnapshot()
  })

  const passwords = [
    { value: 'test', words: [], progress: '0' },
    { value: 'ceci est un test', words: [], progress: '100' },
    { value: 'maison', words: [], progress: '25' },
    { value: 'camion tracteur', words: [], progress: '100' },
    { value: 'camion tracteur2', words: ['camion', 'tracteur2'], progress: '75' },
    { value: 'afeVtj@!#', words: [], progress: '75' }
  ]

  passwords.forEach(password => {
    it(`Test that the component show a progress bar with password ${password.value}`, () => {
      const PasswordInputComponent = shallowMount(PasswordInput, {
        propsData: {
          id: 'passwordInput',
          name: 'passwordInputName',
          label: 'Mon mot de passe',
          value: password.value,
          words: password.words
        }
      })
      expect(PasswordInputComponent.element).toMatchSnapshot()
      expect(PasswordInputComponent.find('vprogresslinear-stub').attributes('value')).toBe(password.progress)
    })
  })
})
