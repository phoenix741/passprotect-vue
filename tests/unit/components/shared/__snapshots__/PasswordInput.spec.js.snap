// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordInput.vue Test that the component show a progress bar 1`] = `
<vtextfield-stub
  append-icon="visibility_off"
  error-messages=""
  hint="This is a top-100 common password"
  id="passwordInput"
  label="Mon mot de passe"
  loading="loading"
  name="passwordInputName"
  persistent-hint="true"
  type="password"
  value="test"
>
  <vprogresslinear-stub
    active="true"
    buffervalue="100"
    color="red accent-2"
    height="3"
    slot="progress"
    value="0"
  />
</vtextfield-stub>
`;

exports[`PasswordInput.vue Test that the component show a progress bar with password afeVtj@!# 1`] = `
<vtextfield-stub
  append-icon="visibility_off"
  error-messages=""
  hint=""
  id="passwordInput"
  label="Mon mot de passe"
  loading="loading"
  name="passwordInputName"
  persistent-hint="true"
  type="password"
  value="afeVtj@!#"
>
  <vprogresslinear-stub
    active="true"
    buffervalue="100"
    color="light-green "
    height="3"
    slot="progress"
    value="75"
  />
</vtextfield-stub>
`;

exports[`PasswordInput.vue Test that the component show a progress bar with password camion tracteur 1`] = `
<vtextfield-stub
  append-icon="visibility_off"
  error-messages=""
  hint=""
  id="passwordInput"
  label="Mon mot de passe"
  loading="loading"
  name="passwordInputName"
  persistent-hint="true"
  type="password"
  value="camion tracteur"
>
  <vprogresslinear-stub
    active="true"
    buffervalue="100"
    color="green"
    height="3"
    slot="progress"
    value="100"
  />
</vtextfield-stub>
`;

exports[`PasswordInput.vue Test that the component show a progress bar with password camion tracteur2 1`] = `
<vtextfield-stub
  append-icon="visibility_off"
  error-messages=""
  hint=""
  id="passwordInput"
  label="Mon mot de passe"
  loading="loading"
  name="passwordInputName"
  persistent-hint="true"
  type="password"
  value="camion tracteur2"
>
  <vprogresslinear-stub
    active="true"
    buffervalue="100"
    color="light-green "
    height="3"
    slot="progress"
    value="75"
  />
</vtextfield-stub>
`;

exports[`PasswordInput.vue Test that the component show a progress bar with password ceci est un test 1`] = `
<vtextfield-stub
  append-icon="visibility_off"
  error-messages=""
  hint=""
  id="passwordInput"
  label="Mon mot de passe"
  loading="loading"
  name="passwordInputName"
  persistent-hint="true"
  type="password"
  value="ceci est un test"
>
  <vprogresslinear-stub
    active="true"
    buffervalue="100"
    color="green"
    height="3"
    slot="progress"
    value="100"
  />
</vtextfield-stub>
`;

exports[`PasswordInput.vue Test that the component show a progress bar with password maison 1`] = `
<vtextfield-stub
  append-icon="visibility_off"
  error-messages=""
  hint="This is a very common password"
  id="passwordInput"
  label="Mon mot de passe"
  loading="loading"
  name="passwordInputName"
  persistent-hint="true"
  type="password"
  value="maison"
>
  <vprogresslinear-stub
    active="true"
    buffervalue="100"
    color="orange"
    height="3"
    slot="progress"
    value="25"
  />
</vtextfield-stub>
`;

exports[`PasswordInput.vue Test that the component show a progress bar with password test 1`] = `
<vtextfield-stub
  append-icon="visibility_off"
  error-messages=""
  hint="This is a top-100 common password"
  id="passwordInput"
  label="Mon mot de passe"
  loading="loading"
  name="passwordInputName"
  persistent-hint="true"
  type="password"
  value="test"
>
  <vprogresslinear-stub
    active="true"
    buffervalue="100"
    color="red accent-2"
    height="3"
    slot="progress"
    value="0"
  />
</vtextfield-stub>
`;
